src/App/Fossa/Lernie/Analyze.hs

```diff
@@ -0,0 +1,223 @@
+analyzeWithLernie ::
+  ( Has Diagnostics sig m
+  , Has Exec sig m
+  , Has ReadFS sig m
+  , Has (Lift IO) sig m
+  ) =>
+  Path Abs Dir ->
+  Maybe ApiOpts ->
+  GrepOptions ->
+  m (Maybe LernieResults)
+analyzeWithLernie rootDir _maybeApiOpts grepOptions = do
+  let maybeLernieConfig = grepOptionsToLernieConfig rootDir grepOptions
+  case maybeLernieConfig of
+    Just (lernieConfig) -> do
+      messages <- runLernie lernieConfig
+      pure $ Just $ lernieMessagesToLernieResults messages rootDir
+    Nothing -> pure Nothing
+
+grepOptionsToLernieConfig :: Path Abs Dir -> GrepOptions -> Maybe LernieConfig
+grepOptionsToLernieConfig rootDir grepOptions =
+  case (customLicenseSearches <> keywordSearches) of
+    [] -> Nothing
+    res -> Just $ LernieConfig rootDir res
+  where
+    customLicenseSearches = map (grepEntryToLernieRegex CustomLicense) (customLicenseSearch grepOptions)
+    keywordSearches = map (grepEntryToLernieRegex KeywordSearch) (keywordSearch grepOptions)
+
+grepEntryToLernieRegex :: LernieScanType -> GrepEntry -> LernieRegex
+grepEntryToLernieRegex scanType grepEntry =
+  LernieRegex (grepEntryMatchCriteria grepEntry) (grepEntryName grepEntry) scanType
+
+runLernie ::
+  ( Has Diagnostics sig m
+  , Has (Lift IO) sig m
+  , Has ReadFS sig m
+  , Has Exec sig m
+  ) =>
+  LernieConfig ->
+  m LernieMessages
+runLernie lernieConfig = withLernieBinary $ \bin -> do
+  let lernieConfigJSON = decodeUtf8 $ Aeson.encode lernieConfig
+  result <- execCurrentDirStdinThrow (lernieCommand bin) lernieConfigJSON
+  pure $ parseLernieJson result
+
+-- Run Lernie, passing "--config -" as its arg so that it gets its config from STDIN
+lernieCommand :: BinaryPaths -> Command
+lernieCommand bin =
+  Command
+    { cmdName = toText $ toPath bin
+    , cmdArgs = ["--config", "-"]
+    , cmdAllowErr = Never
+    }
+
+-- Parse Lernie's NDJson output by splitting on newlines (character 10) and
+-- then decoding each line
+parseLernieJson :: BL.ByteString -> LernieMessages
+parseLernieJson out =
+  fold messages
+  where
+    messageLines = BL.splitWith (== 10) out
+    parsedLines = mapMaybe decode messageLines
+    messages = map singletonLernieMessage parsedLines
+
+lernieMessagesToLernieResults :: LernieMessages -> Path Abs Dir -> LernieResults
+lernieMessagesToLernieResults LernieMessages{..} rootDir =
+  LernieResults
+    { lernieResultsWarnings = lernieMessageWarnings
+    , lernieResultsErrors = lernieMessageErrors
+    , lernieResultsKeywordSearches = keywordSearches
+    , lernieResultsCustomLicenses = customLicenses
+    , lernieResultsSourceUnit = sourceUnit
+    }
+  where
+    keywordSearches = filterLernieMessages lernieMessageMatches KeywordSearch
+    customLicenses = filterLernieMessages lernieMessageMatches CustomLicense
+    sourceUnit = case NE.nonEmpty customLicenses of
+      Nothing -> Nothing
+      Just licenses -> lernieMatchToSourceUnit (NE.toList licenses) rootDir
```

Comment:
[warn, optional] I think this iterates over the list 2 times- it may be worth bucketing messages in one pass with [partition](https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-List.html#v:partition) if you expect the list to be long.